sql/driver/driver_test.go

```diff
@@ -227,3 +229,96 @@ func TestInsecure(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+// Start a transaction to update a row. After updating the row and
+// before committing the transaction, start another transaction in
+// another thread to update the same row. The other transaction succeeds,
+// resulting in the first transaction failing.
+func TestOverlappingTransactions(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	s, db := setup(t)
+	defer cleanup(s, db)
+
+	// Prepare DB
+	if _, err := db.Exec(`CREATE DATABASE t`); err != nil {
+		t.Fatal(err)
+	}
+	if _, err := db.Exec(`CREATE TABLE t.kv (k CHAR PRIMARY KEY, v char)`); err != nil {
+		t.Fatal(err)
+	}
+	if _, err := db.Exec(`INSERT INTO t.kv VALUES ($1, $2)`, "a", "b"); err != nil {
+		t.Fatal(err)
+	}
+
+	// Start first transaction
+	txn, err := db.Begin()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if _, err := txn.Exec(`UPDATE t.kv SET v = $2 WHERE k IN ($1)`, "a", "c"); err != nil {
+		t.Fatal(err)
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	// Run another transaction in parallel that succeeds.
+	go func() {
+		txn2, err := db.Begin()
+		if err != nil {
+			t.Fatal(err)
+		}
+		if rows, err := txn2.Query(`SELECT v FROM t.kv WHERE k IN ($1)`, "a"); err != nil {
+			t.Fatal(err)
+		} else {
+			results := readAll(t, rows)
+			expectedResults := asResultSlice([][]string{
+				{"v"},
+				{"b"},
+			})
+			if err := verifyResults(expectedResults, results); err != nil {
+				t.Fatal(err)
+			}
+		}
+		if _, err := txn2.Exec(`UPDATE t.kv SET v = $2 WHERE k IN ($1)`, "a", "d"); err != nil {
+			log.Fatal(err)
+		}
+		if err := txn2.Commit(); err != nil {
+			log.Fatal(err)
+		}
+		wg.Done()
+	}()
+	// Wait for the other transaction to complete.
+	wg.Wait()
+
+	if rows, err := txn.Query(`SELECT v FROM t.kv WHERE k IN ($1)`, "a"); err != nil {
+		log.Fatal(err)
+	} else {
+		results := readAll(t, rows)
+		// The first update is not visible!
```

Comment:
See "Writer encounters uncommitted write intent" in https://github.com/cockroachdb/cockroach/blob/master/docs/design.md

The second writer upgrades its priority and then succeeds marking the first transaction as aborted.